src/bootstrap/suggest.rs

```diff
@@ -0,0 +1,142 @@
+use crate::builder::Builder;
+use glob::Pattern;
+use std::fmt::{self, Display};
+
+/// Suggests a list of possible `x.py` commands to run based on modified files in branch.
+pub fn suggest(build: &Builder<'_>, run: bool) {
+    // Find the commit the current branch split from master.
+    let merge_base = build
+        .config
+        .git()
+        .arg("merge-base")
+        .arg("HEAD")
+        .arg("origin/master")
+        .output()
+        .expect("failed to execute `git`");
+
+    // `git` shouldn't ever give us non-utf8 output.
+    let merge_base = String::from_utf8(merge_base.stdout).expect("`git` gave non-utf8 commit?!");
+
+    // Note that `merge_base` ends in a newline; it must be trimmed.
+    let modified_files = build
+        .config
+        .git()
+        .arg("diff")
+        .arg("--name-only")
+        .arg(merge_base.trim())
+        .output()
+        .expect("failed to execute `git`");
+
+    let modified_files =
+        String::from_utf8(modified_files.stdout).expect("`git` gave non-utf8 commit?!");
+
+    let modified_files: Vec<_> = modified_files.split_terminator('\n').collect();
+
+    println!("=== MODIFIED FILES ===");
+    for file in &modified_files {
+        println!("{}", file);
+    }
+    println!("======================\n");
+
+    let mut suggestions = Vec::new();
+    for sug in SUGGESTIONS {
+        let glob = Pattern::new(sug.0).expect("Found invalid glob pattern!");
+
+        for file in &modified_files {
+            if glob.matches(file) {
+                suggestions.extend_from_slice(sug.1);
+            }
+        }
+    }
+
+    suggestions.sort();
+    suggestions.dedup();
+
+    println!("=== SUGGESTIONS ===");
+    for sug in &suggestions {
+        println!("{}", sug);
+    }
+    println!("===================");
+
+    if run {
+        for _sug in suggestions {
+            // HELP!: how should I use `rustbuild`'s internal plumbing to achieve this?
+        }
+    } else {
+        println!("help: consider using the `--run` flag to automatically run suggested tests");
+    }
+}
+
+#[derive(Clone, PartialOrd, Ord, PartialEq, Eq)]
+struct Suggestion {
+    subcommand: &'static str,
+    stage: Option<u8>,
+    paths: &'static [&'static str],
+}
+
+impl Display for Suggestion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(f, "x.py {} ", self.subcommand)?;
+
+        if let Some(stage) = self.stage {
+            write!(f, "--stage {} ", stage)?;
+        }
+
+        for path in self.paths {
+            write!(f, "{} ", path)?;
+        }
```

Comment:
Hmm, this reminds me a lot of https://github.com/rust-lang/rust/issues/97338. I wonder if we should use that approach here, have steps tell you have to rerun them. it's a *lot* of work unfortunately :/ but it seems unfortunate to duplicate it.

on the other hand that would couple this pretty tightly to bootstrap which I don't want to do; so ignore me, sorry.